plugins/api/src/switch_user.cpp

```diff
@@ -0,0 +1,253 @@
+                return;
+            }
+
+            // Disconnect if the remote server's version is older than 4.3.1.
+            // Remember, only iRODS 4.3.1 and later support the switch user API plugin.
+            if (*irods::to_version(_host.conn->svrVersion->relVersion) < irods::version{4, 3, 1}) {
+                rcDisconnect(_host.conn);
+                _host.conn = nullptr;
+                return;
+            }
+
+            // At this point, we know the remote server supports this API plugin.
+            if (const auto ec = rc_switch_user(_host.conn, client.userName, client.rodsZone); ec != 0) {
+                log_api::error(
+                    "rc_switch_user failed on remote host [{}] with error_code [{}]. Disconnecting from remote host.",
+                    _host.hostName->name,
+                    ec);
+                rcDisconnect(_host.conn);
+                _host.conn = nullptr;
+            }
+        };
+
+        for (auto* zone_ptr = ZoneInfoHead; zone_ptr; zone_ptr = zone_ptr->next) {
+            for (auto* host_ptr = zone_ptr->primaryServerHost; host_ptr; host_ptr = host_ptr->next) {
+                switch_user_or_disconnect(*host_ptr);
+            }
+
+            for (auto* host_ptr = zone_ptr->secondaryServerHost; host_ptr; host_ptr = host_ptr->next) {
+                switch_user_or_disconnect(*host_ptr);
+            }
+        }
+    } // update_server_to_server_connections
+
+    auto rs_switch_user(RsComm* _comm, SwitchUserInput* _input) -> int
+    {
+        try {
+            // Only administrators are allowed to invoke this API.
+            // We check the proxy user because it covers clients and server-to-server redirects.
+            // Remember, the client and proxy user information is identical until there's a redirect.
+            if (!irods::is_privileged_proxy(*_comm)) {
+                log_api::error(
+                    "Proxy user [{}] does not have permission to switch client user.", _comm->proxyUser.userName);
+                return SYS_PROXYUSER_NO_PRIV;
+            }
+
+            // Return immediately if the client did not provide non-empty strings for the
+            // username and zone.
+            if (const auto ec = check_input(_input); ec != 0) {
+                return ec;
+            }
+
+            const auto* local_zone = getLocalZoneName();
+
+            if (!local_zone) {
+                log_api::error("Could not get name of local zone.");
+                return SYS_INTERNAL_ERR;
+            }
+
+            namespace adm = irods::experimental::administration;
+
+            // This call covers the existence check too.
+            const auto user_type = adm::server::type(*_comm, adm::user{_input->username, _input->zone});
+
+            if (!user_type) {
+                log_api::error(
+                    "[{}#{}] is not a user in the local zone [{}].", _input->username, _input->zone, local_zone);
```

Comment:
This wording makes me think that remote users are not allowed or considered here. Maybe this could say:
```
User type for [{}#{}] could not be identified.
```
Wording is fine as is, just offering this for clarity

Reply:
Will consider it.

Reply:
How does this sound?
```
User type for [{}#{}] could not be identified in local zone [{}].
```

Reply:
that seems good and specific/thorough

Reply:
done